fix(price-impact): improve price impact loading

alfetopito opened the pull request:
# Summary

Improve price impact loading

## Notes
- So far only really checked SWAP
- LIMIT orders might still not work as expected - working on it (either this PR or a follow up)
- There's a lot of noise as the fiat hooks interface changed to propagate the loading indicator flag

# To Test

1. On mainnet, pick sell and buy tokens and fill in the sell input
* Price impact will be calculated
* There should be a loading indicator to the right of the price impact
2. Change the amounts
* Price impact should be updated accordingly
* Loading indicator should be seen

alfetopito commented on src/common/hooks/useSafeMemo.ts:
```diff
@@ -21,9 +22,13 @@ export function useSafeMemoDeps(deps: unknown[]): unknown[] {
   return useSafeMemo<typeof depsObj>(() => depsObj, Object.values(depsObj))
 }
+
+export function useSafeEffect(memoCall: EffectCallback, deps: unknown[]): void {
+  useEffect(memoCall, useSafeDeps(deps))
+}
```
New `Safe` hook

alfetopito commented on src/legacy/hooks/usePriceImpact/useFiatValuePriceImpact.ts:
```diff
@@ -5,9 +5,9 @@ import { computeFiatValuePriceImpact } from 'legacy/utils/computeFiatValuePriceI
 export default function useFiatValuePriceImpact(parsedAmounts: ParsedAmounts) {
-  const fiatValueInput = useHigherUSDValue(parsedAmounts[Field.INPUT])
-  const fiatValueOutput = useHigherUSDValue(parsedAmounts[Field.OUTPUT])
+  const { value: fiatValueInput, isLoading: inputIsLoading } = useHigherUSDValue(parsedAmounts[Field.INPUT])
+  const { value: fiatValueOutput, isLoading: outputIsLoading } = useHigherUSDValue(parsedAmounts[Field.OUTPUT])
   const priceImpact = computeFiatValuePriceImpact(fiatValueInput, fiatValueOutput)
 
-  return priceImpact
+  return { priceImpact, isLoading: inputIsLoading || outputIsLoading }
```
`isLoading` flag being propagated

alfetopito commented on src/modules/trade/updaters/PriceImpactUpdater.tsx:
```diff
@@ -1,21 +1,20 @@
-    updatePriceImpact(priceImpact)
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [priceImpactValue, priceImpact.loading, priceImpact.error, updatePriceImpact])
+  useSafeEffect(() => {
+    updatePriceImpact({ error, loading, priceImpact })
+  }, [error, loading, updatePriceImpact, priceImpact])
```
This was not behaving very well, thus the new `Safe` hook

alfetopito commented on src/modules/trade/hooks/useTradePriceImpactParams.ts:
```diff
@@ -34,6 +34,5 @@ export function useTradePriceImpactParams(): PriceImpactParams {
     }
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [inputCurrencyAmount?.quotient.toString(), outputCurrencyAmount?.quotient.toString(), isWrapOrUnwrap])
+  }, [inputCurrencyAmount, outputCurrencyAmount, isWrapOrUnwrap])
```
This was also behaving weirdly, replaced with the `Safe` version

alfetopito commented on src/common/hooks/useSafeMemo.ts:
```diff
@@ -1,15 +1,16 @@
     if (dep instanceof Token) return dep.address.toLowerCase()
     if (dep instanceof CurrencyAmount) return dep.toExact() + dep.currency.symbol + dep.currency.isNative
     if (dep instanceof WrappedTokenInfo) return dep.address
+    if (dep instanceof Percent) return dep.toFixed(6)
```
Is this enough to check it's different, right?

Actually, now that I'm seeing this I'm considering making it generic for any fraction with https://github.com/cowprotocol/cowswap/blob/0ae6602f707045079c8a5e0f3b8d6bd70f703e61/src/utils/fractionUtils.ts#L12-L11

shoom3301 commented:
<img width="507" alt="image" src="https://github.com/cowprotocol/cowswap/assets/7122625/e5f3bd57-b61f-42ab-9bc4-2d1d25f4c898">

@alfetopito I've noticed one artifact: the loader indicator is displayed in the output token input event when there is not currency selected

shoom3301 commented:
<img width="511" alt="image" src="https://github.com/cowprotocol/cowswap/assets/7122625/457f75c1-1b22-4ab3-bfc0-78fa0bd99d1c">

https://github.com/cowprotocol/cowswap/assets/7122625/94b4c510-b72a-4926-911c-3b1e4173667a

@alfetopito another nuance with the loading indicator:
It is displayed at the same time with the price impact value. In dev we display only something one at one moment of the time.

shoom3301 reviewed (changes requested):
Need to fix a couple of issues with the loading indicator conditions

elena-zh reviewed (changes requested):
Hey @alfetopito , I will add a bit more cases when price impact loader is displayed (and no needed): 
1. Not connected wallet 
![not connected](https://github.com/cowprotocol/cowswap/assets/70885163/08d3ca1f-20b6-4946-a703-39550b091a82)
2. Unsupported token 
![unsupp](https://github.com/cowprotocol/cowswap/assets/70885163/57bdfc06-58c6-42af-bfd5-3d0effc0243e)
3. Sell amount is too small in Advanced orders page 
![twap sell amount too small](https://github.com/cowprotocol/cowswap/assets/70885163/2063bfbb-e88d-445a-a1d5-18032061e619)
4. Unsupported network 
![uns chain](https://github.com/cowprotocol/cowswap/assets/70885163/167f020d-134e-4995-a243-5c06a78904dc)
5. Impossible to calculate a price impact 
![not possible to calculate](https://github.com/cowprotocol/cowswap/assets/70885163/4a5db6d8-9dc1-4132-9d25-fb034189f6fa)
6. Select WRAP/UWRAP TX on Swap page, then switch to Limit/Advanced orders tabs: there will be a loader 
![limit unwrap](https://github.com/cowprotocol/cowswap/assets/70885163/a102b7f4-7690-4622-a2de-2a84db15781b)
![twap unwrap](https://github.com/cowprotocol/cowswap/assets/70885163/dbad50d4-3db0-41c5-89d8-22174b82a21b)

Thanks!

alfetopito commented:
Hmmm, that might indicate that we are loading a lot of times when we shouldn't.
Will investigate, thanks for the input so far!

alfetopito commented:
@shoom3301 @elena-zh it should be fixed now

Turns out we were (since forever?) always querying the fiat value even if there was only 1 token selected.
Now it will only trigger that when both are selected and there is an input amount.
Also not calculating when it's a wrap and so on

alfetopito commented on src/legacy/hooks/usePriceImpact/useFallbackPriceImpact.ts:
```diff
@@ -55,7 +55,8 @@ export default function useFallbackPriceImpact({
   // Should we even calc this? Check if fiatPriceImpact exists OR user is wrapping token
-  const shouldCalculate = !!abTrade && !isWrapping && isWindowVisible
+  const shouldCalculate =
+    !!abTrade && !isWrapping && isWindowVisible && !!abTrade?.inputAmount && !!abTrade?.outputAmount
```
Skip fallback impact calculations unless both input and output are set

alfetopito commented on src/legacy/hooks/usePriceImpact/useFiatValuePriceImpact.ts:
```diff
@@ -1,13 +1,19 @@
-  const fiatValueInput = useHigherUSDValue(parsedAmounts[Field.INPUT])
-  const fiatValueOutput = useHigherUSDValue(parsedAmounts[Field.OUTPUT])
+export default function useFiatValuePriceImpact({ INPUT, OUTPUT }: ParsedAmounts) {
+  const areBothValuesPresent = !!INPUT && !!OUTPUT
+  // prevent querying any fiat estimation unless both values are filled in
+  const input = areBothValuesPresent ? INPUT : undefined
+  const output = areBothValuesPresent ? OUTPUT : undefined
+
+  const { value: fiatValueInput, isLoading: inputIsLoading } = useHigherUSDValue(input)
+  const { value: fiatValueOutput, isLoading: outputIsLoading } = useHigherUSDValue(output)
   const priceImpact = computeFiatValuePriceImpact(fiatValueInput, fiatValueOutput)
 
-  return priceImpact
+  const isLoading = areBothValuesPresent && (inputIsLoading || outputIsLoading)
+
+  return { priceImpact, isLoading }
```
Skip fiat impact calculations unless both input and output are set

elena-zh commented:
Hey @alfetopito , the last nitpick is that price impact loader is still displayed when 'sell amount is too small' error is displayed on Limit orders page
![sell amount is too small](https://github.com/cowprotocol/cowswap/assets/70885163/af496045-3b33-4748-99d5-bfcaa76daf78)

alfetopito commented:
> Hey @alfetopito , the last nitpick is that price impact loader is still displayed when 'sell amount is too small' error is displayed on Limit orders page ![sell amount is too small](https://user-images.githubusercontent.com/70885163/258802662-af496045-3b33-4748-99d5-bfcaa76daf78.png)

That is a different problem and I actually don't think it's a problem.
It's it too bad to have the estimation in that case?
If it is, I'll can work in a different PR

elena-zh commented:
@alfetopito , I can't say that it is too bad. The thing is that it works differently than on other forms. 
Then, I see it loading almost every time. 
I can open a separate task for this if needed :)

alfetopito commented:
> @alfetopito , I can't say that it is too bad. The thing is that it works differently than on other forms. Then, I see it loading almost every time. I can open a separate task for this if needed :)

Oh, didn't notice it was different only on Limit form.
I'll access if it's a quick fix to go in this PR or whether it should be as a separated task.

shoom3301 commented:
<img width="693" alt="image" src="https://github.com/cowprotocol/cowswap/assets/7122625/afd7577b-2cc4-4c07-a452-e5c45c5aff8a">

@alfetopito I still see the fiat value + price impact + loading indicator at the same time.
Hot to reproduce:
1. On Mainnet: 50000 COW > WETH
2. Changed buy token to WBTC
3. Changed buy token to WETH
4. Repeat 2-3 steps

alfetopito commented:
> <img alt="image" width="693" src="https://user-images.githubusercontent.com/7122625/259013039-afd7577b-2cc4-4c07-a452-e5c45c5aff8a.png">
> 
> @alfetopito I still see the fiat value + price impact + loading indicator at the same time. Hot to reproduce:
> 
>     1. On Mainnet: 50000 COW > WETH
> 
>     2. Changed buy token to WBTC
> 
>     3. Changed buy token to WETH
> 
>     4. Repeat 2-3 steps

@shoom3301 should be fixed now 🤞 

@elena-zh decided not to hide it as the fiat estimation (and price impact) are bound to the amounts picked, so I think it's not a problem to have them here.
(Also because took me quite awhile and I did not manage to find a good way to hide it so I gave up 🙈)

elena-zh reviewed (approved):
LGTM then!
At least, I don't see eternal loader when I see price impact for 'sell amount to small' case on the Limit orders page :)

shoom3301 reviewed (approved):
Works great!
